### ISSUE TITLE
[Bug] Bincode deserialization of tuples containing enums is broken in c++

### ISSUE THREAD

**Issue by weiznich** [zefchain/serde-reflection#31]
# 🐛 Bug

## To reproduce

**Code snippet to reproduce**
```rust

#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Hash, Serialize, Deserialize)]
pub enum Key {
    Id,
    Custom(String),
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Value {
    NotSet,
    Integer(i32),
    String(String),
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MyStruct {
    pub a: (Key, Value),
    pub o: Key,
}

fn make_attribute() -> MyStruct {
    MyStruct {
        a: (Key::Custom("abc".into()), Value::String("bar".into())),
        o: Key::Id,
    }
}

#[no_mangle]
pub unsafe extern "C" fn test_in(data: *const u8, len: usize) {
    if !data.is_null() {
        let slice = std::slice::from_raw_parts(data, len);
        let data = bincode::deserialize::<MyStruct>(slice).unwrap();
        assert_eq!(data, make_attribute());
    }
}

#[no_mangle]
pub unsafe extern "C" fn test_out(data: *mut *mut u8) -> usize {
    let p = make_attribute();

    let binary = bincode::serialize(&p).unwrap();
    let b = binary.into_boxed_slice();
    let len = b.len();
    let out: &mut *mut u8 = &mut *data;
    *out = Box::into_raw(b) as *mut u8;
    len
}

```

```c++
#include <iostream>
#include "include/testing.hpp"

using namespace std;

extern "C" size_t test_out(unsigned char** out_ptr);
extern "C" void test_in(const unsigned char* data, size_t len);

int main(int argc, char *argv[]) {


    uint8_t* out_ptr1 = nullptr;
    size_t len1 = test_out(&out_ptr1);
    test_in(out_ptr1, len1);
    std::cout << "Just in and out works fine" << std::endl;

    uint8_t* out_ptr = nullptr;
    size_t len = test_out(&out_ptr);

    std::vector<uint8_t> serialized_result(out_ptr, out_ptr + len);
    auto res = testing::MyStruct::bincodeDeserialize(serialized_result);

    auto serializer = serde::BincodeSerializer();
    serde::Serializable<testing::MyStruct>::serialize(res, serializer);
    std::vector<uint8_t> bytes = std::move(serializer).bytes();

    test_in(bytes.data(), bytes.size());

    std::cout << "Everything is fine" << std::endl;

    return 0;
}

```

(I'm aware that this leaks memory and unwinds through the ffi boundary.)

**Stack trace/error message**
```
terminate called after throwing an instance of 'serde::deserialization_error'
  what():  Unknown variant index for enum
```

See [here](https://github.com/weiznich/serde_generate_issue) for an example project reproducing the error. Assuming a linux environment with g++ installed you only need to clone that repository and execute `make` to run all required steps to see this error message

## Expected Behavior

I would expect that the c++ code executes without throwing an exception.  That would imply that it prints the following messages:
```
Just in and out works fine
Everything is fine
```

## System information

**Please complete the following information:**
`serde-reflection`: `0.3.6`
`serde-generate`: `0.25.0`

`rustc`: `rustc 1.67.0 (fc594f156 2023-01-24)`

## Additional context

Slightly changing the layout of the `MyStruct` type and/or the `Value` enum changes the generated error message. For example removing [this line](https://github.com/weiznich/serde_generate_issue/blob/1c0525de7049904a4c97ad0d5122e7fb42f5fb88/src/lib.rs#L13) removes the c++ exception but causes [this](https://github.com/weiznich/serde_generate_issue/blob/1c0525de7049904a4c97ad0d5122e7fb42f5fb88/src/lib.rs#L35) assert to fail with the following message:
```
thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
  left: `MyStruct { a: (Custom("bar"), String("abc")), o: Id }`,
 right: `MyStruct { a: (Custom("abc"), String("bar")), o: Id }`', src/lib.rs:35:9
```
Which implies that there is something wrong with how enum variants are handled inside of a tuple.

**Comment by weiznich** [comment]
@ma2bd Any ideas how to fix that?

**Comment by ma2bd** [comment]
Oh I just saw this. Let me have a look

**Comment by ma2bd** [comment]
Thanks @weiznich for the high quality bug report. I cloned your [project](https://github.com/weiznich/serde_generate_issue) and running `make` actually works fine on my Macbook, with g++ = "Apple clang version 13.1.6".

I'm a bit puzzled. Let me try to add a unit test for the Linux CI.

**Comment by ma2bd** [comment]
oops didn't mean to close the issue just yet

**Comment by ma2bd** [comment]
Once we have a clean repro, I suspect the issue is related to https://stackoverflow.com/questions/43959229/does-stdapply-provide-a-guarantee-for-order-of-evaluation

**Comment by ma2bd** [comment]
FWIW using `g++-12` from homebrew's gcc also works on my laptop.

**Comment by weiznich** [comment]
Given the fact that it seems to switch key/value with the following patch to the test case:

```diff
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -10,20 +10,20 @@ pub enum Key {
 #[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
 pub enum Value {
     NotSet,
-    Integer(i32),
+//    Integer(i32),
     String(String),
 }
 
 #[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
 pub struct MyStruct {
     pub a: (Key, Value),
-    pub o: Key,
+    //pub o: Key,
 }
 
 fn make_attribute() -> MyStruct {
     MyStruct {
         a: (Key::Custom("abc".into()), Value::String("bar".into())),
-        o: Key::Id,
+        //o: Key::Id,
     }
 }

```

results in
```
thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
  left: `MyStruct { a: (Custom("bar"), String("abc")) }`,
 right: `MyStruct { a: (Custom("abc"), String("bar")) }`', src/lib.rs:35:9
```

I guess that the issue might be here: https://github.com/zefchain/serde-reflection/blob/7facc0412d060a7cf20d1da17d75b014b1218fab/serde-generate/runtime/cpp/serde.hpp#L613 ?

**Comment by weiznich** [comment]
@ma2bd I've opened #32 to fix this.

**Comment by ma2bd** [comment]
@weiznich Thanks for the fix! I suspected an issue at this line but couldn't find the right explanation.

**Comment by ma2bd** [comment]
Released in serde-generate v0.25.1
